Optimize count queries for fixed cardinality

navinvishy opened the pull request:
Fixes #7628

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestSimplifyCountOverInnerAggregateQuery.java:
```diff
@@ -0,0 +1,136 @@
+                                            ImmutableList.of(totalprice),
+                                            ImmutableMap.of(totalprice, new TpchColumnHandle(totalprice.getName(), DOUBLE)))),
+                            ImmutableMap.of(p.symbol("sum_inner", DOUBLE),
+                                    new AggregationNode.Aggregation(
```
why not to use `p.aggregation(...)` here and in all the places where you create a new aggregation node?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyCountOverInnerAggregateQuery.java:
```diff
@@ -0,0 +1,69 @@
+            AggregationNode.Aggregation aggregation = entry.getValue();
+            if (aggregation != null) {
+                Signature signature = aggregation.getSignature();
+                if (!"count".equals(signature.getName())) {
```
I think you need check also if `count` has arguments as `count(*)` has different semantic than `count(a)`

findepi commented:
@navinvishy, @electrum, what are the real-life use-cases for this?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyCountOverInnerAggregateQuery.java:
```diff
@@ -0,0 +1,69 @@
+        if (!(child instanceof AggregationNode)) {
+            return Optional.empty();
+        }
+        AggregationNode childAggregationNode = (AggregationNode) child;
```
Instead of checking nested aggregation you can use `ScalarQueryUtil`

navinvishy commented:
@kokosing Thanks for the feedback! I've made the changes you suggested and added a couple of test cases to TestLogicalPlanner. Could you please take another look.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:
```diff
@@ -280,7 +281,9 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                         new AddIntermediateAggregations(),
                         new RemoveRedundantIdentityProjections())));
-
+        builder.add(new IterativeOptimizer(
```
Merge your rule with the rules above. There is no point of separating them.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyCountOverInnerAggregateQuery.java:
```diff
@@ -60,8 +63,7 @@
         }
-        AggregationNode childAggregationNode = (AggregationNode) child;
-        if (childAggregationNode.getAssignments().isEmpty() && childAggregationNode.hasEmptyGroupingSet() && !childAggregationNode.hasNonEmptyGroupingSet()) {
+        if (ScalarQueryUtil.isScalar(child, lookup)) {
```
Once you used this method, your rule become more generic as it can handle source like EnsureSingleRow or Values with single row, so you need to change the rule name.

navinvishy replied:
@kokosing good point, thanks. I've made these changes. Good for another look.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -370,4 +375,28 @@ public void testCorrelatedScalarAggregationRewriteToLeftOuterJoin()
+        assertPlan(
+                "SELECT count(*) from (select sum(orderkey) from orders)",
+                output(node(ValuesNode.class).with(new ValuesMatcher(ImmutableMap.of(), Optional.of(Integer.valueOf(1)), Optional.of(ImmutableList.of(ImmutableList.of(new LongLiteral("1"))))))));
+    }
```
put a single line below

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -370,4 +375,28 @@ public void testCorrelatedScalarAggregationRewriteToLeftOuterJoin()
+    {
+        assertPlan(
+                "SELECT count(*) from (select sum(orderkey) from orders)",
+                output(node(ValuesNode.class).with(new ValuesMatcher(ImmutableMap.of(), Optional.of(Integer.valueOf(1)), Optional.of(ImmutableList.of(ImmutableList.of(new LongLiteral("1"))))))));
```
use a `com.facebook.presto.sql.planner.assertions.PlanMatchPattern#values` here. Please put each node matcher in separate line

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -370,4 +375,28 @@ public void testCorrelatedScalarAggregationRewriteToLeftOuterJoin()
+                            aggregation(ImmutableMap.of("FINAL_SUM", functionCall("sum", ImmutableList.of("PARTIAL_SUM"))),
+                                    exchange(ExchangeNode.Scope.LOCAL, ExchangeNode.Type.GATHER,
+                                            aggregation(ImmutableMap.of("PARTIAL_SUM", functionCall("sum", ImmutableList.of("ORDERKEY"))),
+                                                    tableScan("orders", ImmutableMap.of("ORDERKEY", "orderkey"))
```
I would just use anyTreeNode with tableScan as child. That would mean exactly that plan was not optimized.

```
 assertPlan(
                 "SELECT count(s) from (select sum(orderkey) as s from orders)",
                 anyTreeNode(
                     tableScan("orders')));
```

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyCountOverScalar.java:
```diff
@@ -0,0 +1,71 @@
+ * A count over a subquery can be reduced to a VALUES(1) provided
+ * the subquery is a scalar
+ */
+public class SimplifyCountOverScalar
```
I would name it `PruneCountAggregationOveScalar`, what do you think?

navinvishy replied:
@kokosing not sure if PruneCountAggregationOveScalar or TransformCountAggregationOveScalar is better (since we're not really pruning). I've named it Prune* for now anyway...

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyCountOverScalar.java:
```diff
@@ -0,0 +1,71 @@
+        }
+        for (Map.Entry<Symbol, AggregationNode.Aggregation> entry : assignments.entrySet()) {
+            AggregationNode.Aggregation aggregation = entry.getValue();
+            if (aggregation != null) {
```
use requireNonNull, IMO null in this map suggests some kind of bug

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyCountOverScalar.java:
```diff
@@ -0,0 +1,71 @@
+        }
+        AggregationNode parent = (AggregationNode) node;
+        Map<Symbol, AggregationNode.Aggregation> assignments = parent.getAggregations();
+        if (assignments.size() > 1) {
```
what if size is 0? IMO you should only rewrite plan if there is only a single aggregation function

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -370,4 +373,22 @@ public void testCorrelatedScalarAggregationRewriteToLeftOuterJoin()
+    }
+
+    @Test
+    public void testCountColumnOverInnerAggregateQuery()
```
I would merge this two methods together and mention which rule you are testing in the method name

kokosing commented:
Also please implement `Rule#getPattern` method

navinvishy commented:
@kokosing in case you havent seen it yet, I've addressed your comments. Please take another look.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCountAggregationOverScalar.java:
```diff
@@ -0,0 +1,81 @@
+            }
+        }
+        PlanNode child = lookup.resolve(parent.getSource());
+        if (!(child instanceof AggregationNode)) {
```
if you are using `ScalarQueryUtil` you don't need this `if` block

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:
```diff
@@ -279,8 +280,8 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
-                        new RemoveRedundantIdentityProjections())));
-
+                        new RemoveRedundantIdentityProjections(),
+                        new PruneCountAggregationOverScalar())));
```
why is it added so late? I would add it add line 150 or so. The sooner this optimization is fired the better for other rules.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCountAggregationOverScalar.java:
```diff
@@ -0,0 +1,129 @@
+
+public class TestPruneCountAggregationOverScalar extends BasePlanTest
+{
+    private RuleTester tester;
```
add a blank line after this line

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCountAggregationOverScalar.java:
```diff
@@ -0,0 +1,129 @@
+                                        .groupingSets(ImmutableList.of(ImmutableList.of()))
+                                        .step(AggregationNode.Step.SINGLE)
+                )).matches(PlanMatchPattern.values(ImmutableMap.of("count_1", 0)));
+    }
```
add blank line

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCountAggregationOverScalar.java:
```diff
@@ -0,0 +1,129 @@
+                .on(p ->
+                        p.aggregation((a) -> a.addAggregation(
+                                p.symbol("count_1", BigintType.BIGINT),
+                                new AggregationNode.Aggregation(
```
[optional]

I do not like the fact you need to call constructors of `PlanNode` internals directly, it would be nice to have a better support for creating aggregation nodes in `PlanBuilder`

Can you please take a look and check if there is nothing there already which you could use? If there is nothing helpful there, maybe you could add something.

In current form this difficult to understand actual shape of plan as it is obscured with all the boilerplate like `ImmutableList.of()`.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneCountAggregationOverScalar.java:
```diff
@@ -0,0 +1,129 @@
+import static com.facebook.presto.tpch.TpchMetadata.TINY_SCALE_FACTOR;
+import static io.airlift.testing.Closeables.closeAllRuntimeException;
+
+public class TestPruneCountAggregationOverScalar extends BasePlanTest
```
please add tests that your rule is fired on aggregation over single row `Values` and `EnforceSingleRow`

kokosing replied:
should not be `extends` be in separate line?

kokosing replied:
Add tests where nested aggregation has non empty `GROUP BY` clause

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCountAggregationOverScalar.java:
```diff
@@ -0,0 +1,81 @@
+        }
+        AggregationNode parent = (AggregationNode) node;
+        Map<Symbol, AggregationNode.Aggregation> assignments = parent.getAggregations();
+        if (assignments.size() == 0 || assignments.size() > 1) {
```
you need to also check that there no grouping sets, I mean there is one default.

See com.facebook.presto.sql.planner.plan.AggregationNode#hasDefaultOutput

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -370,4 +373,17 @@ public void testCorrelatedScalarAggregationRewriteToLeftOuterJoin()
     }
+
+    @Test
+    public void testPruneCountAggregationOverScalar()
```
Please also test following queries:

SELECT count(*) from (select sum(orderkey) from orders GROUP  BY customerkey)
SELECT EXISTS(SELECT sum(orderkey) from orders)

With the rule you are adding, the second query should also be optimized to VALUES

navinvishy replied:
@kokosing the second query gets optimized to:

```
- Output[_col0] => [expr_4:boolean]
        Cost: {rows: 1, bytes: ?}
        _col0 := expr_4
    - Project[] => [expr_4:boolean]
            Cost: {rows: 1, bytes: ?}
            expr_4 := true
        - Values => []
                Cost: {rows: 1, bytes: ?}
                ()
```
But this rule applies only to aggregations (getPattern() matches AggregationNode). This rule could be modified to apply to this query but I feel like that should be in different rule. What do you think?

kokosing replied:
`EXISTS` is rewritten to aggregation, and then it is optimized with your optimization.

The first query should not be optimized.

navinvishy replied:
@kokosing could you please point me to where and when in the code this rewrite happens? I got as far as RelationPlanner and QueryPlanner but I dont see where an EXISTS is getting rewritten to an AggregationNode. I'm assuming there is some AstVisitor that does this...

navinvishy replied:
Also, I noticed that in LogicalPlanner.plan(), even before applying any optimizers, the AST looks to have been rewritten to an OutputNode followed by a series of ProjectNode children. Not sure when exactly this optimizer will see a tree containing an AggregationNode.

kokosing replied:
com.facebook.presto.sql.planner.iterative.rule.TransformExistsApplyToLateralNode

kokosing replied:
Then lateral can be rewritten to join with com.facebook.presto.sql.planner.optimizations.TransformUncorrelatedLateralToJoin

navinvishy replied:
@kokosing I've addressed all your comments. The thing I see is that for the `SELECT EXISTS(SELECT sum(orderkey) from orders)` query, the plan gets rewritten to a scalar value even before optimization. I believe that happens here: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java#L284.
If I understand this correctly, the rule I'm adding will not see an Exists node because of this rewrite. Please let me know if that is correct.

kokosing replied:
good point.  Now with your rewrite that `if` you pointed should be no longer needed.

kokosing replied:
ok, I think it was good that you investigated this. Thank you. Once we finish this PR, in some spare time I will try to remove that `if` (https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java#L284.).

Lets put `EXISTS` aside of this PR.

kokosing commented:
Please also make sure that affected queries are covered in AbstractTestQueries.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCountAggregationOverScalar.java:
```diff
@@ -0,0 +1,78 @@
+                return Optional.empty();
+            }
+        }
+        PlanNode child = lookup.resolve(parent.getSource());
```
IMO `isScalar` will take care of resolving nodes, so you do not need to resolve it here

kokosing commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -8933,4 +8933,21 @@ public void testLateralJoin()
+    @Test
+    public void testCountAggregationOverScalar()
+    {
+        MaterializedResult actual = computeActual(
```
why not to use plain `assertQuery`, then whole code of this method can be rewritten to:
```
assertQuery("SELECT COUNT(*) FROM (SELECT SUM(orderkey) FROM orders)");
```

navinvishy commented:
@kokosing I've addressed your comments. I see the travis build is failing but I'm not sure its related to these changes. One stacktrace had this at the bottom: ``` Caused by: java.net.UnknownHostException: presto-master.docker.cluster```

kokosing commented:
@navinvishy I applied following changes on your patch, once travis passes I will merge this.

https://gist.github.com/a9726c9a0f681efae630bdfce84ee3a6

kokosing commented:
Thanks!
